### PR TITLE
Jenkins job to create auth0 passwordless user, and permission/role/group

### PR DESCRIPTION
**Pull request by xoen**
This script create the passwordless user (`"email"` connection) in Auth0.
Then it creates the permission (`"view:app"`'), role (`"app-viewer"`) and group (same name as app/client) in the Authorization Extension. These are created only if they don't already exist. Also, existing permissions/roles/members are left alone when adding them to role/group/group.

**NOTE**: We may decide to move the creation of permission/role/group at the time of build of the shiny app to avoid waste of API calls.

### CODE REVIEW

**Comment by xoen** [comment]
@andyhd probably far from perfect as first Python code but it works.

**Comment by xoen** [comment]
Also, as I said, potentially the creation of permission/role/group could be moved. That's one of the reason for extracting the logic in a classes. May be too early for a full fledge python module to install with pip, etc...

**Comment by xoen** [comment]
@andyhd FYI: also added jenkins job: https://github.com/ministryofjustice/analytics-platform-ops/pull/35/commits/1293967b13eeff8d254197a3511e2983a21cf408

File: jenkinsfiles/auth0-users/role.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+    def add_permission(self, permission_id):
+        payload = {
+            'name': self.name(),
+            'description': self.description(),
+            'applicationId': self.application_id(),
+            'applicationType': self.application_type(),
+            'permissions': self.permission_ids(),
+        }
```
Comment (andyhd) [inline jenkinsfiles/auth0-users/role.py]: Can `payload` not just be `self.data` here?

File: jenkinsfiles/auth0-users/role_api.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+from role import Role
+
+import requests
+import logging
```
Comment (andyhd) [inline jenkinsfiles/auth0-users/role_api.py]: Import order and grouping

File: jenkinsfiles/auth0-users/role_api.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+logging.basicConfig()
+LOG = logging.getLogger(__name__)
+LOG.setLevel(logging.DEBUG)
+
+
+class RoleAPI(object):
+
+    def __init__(self, authz_api, authz_token):
+        self.authz_api = authz_api
+        self.authz_token = authz_token
```
Comment (andyhd) [inline jenkinsfiles/auth0-users/role_api.py]: All these API classes have a lot in common - maybe there should be a superclass?

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('domain', help="Auth0 domain")
+    parser.add_argument('client_id', help="Auth0 client ID")
+    parser.add_argument('client_secret', help="Auth0 client secret")
+    parser.add_argument('authz_api', help="Authorization API URL")
+    parser.add_argument('app_name', help="App client name")
+    parser.add_argument('email', help="User email")
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: Probably not in this case.

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            LOG.debug("App found = {}".format(app))
+            return app
+    LOG.critical("App with name '{}' not found".format(app_name))
+
+
+def get_authz_token(domain, client_id, client_secret):
+    get_token = GetToken(domain)
+    credentials = get_token.client_credentials(
+        client_id, client_secret, 'urn:auth0-authz-api'
+    )
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: Very well spotted. For some reason I didn't think about it. Thanks for the suggestion!

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    return credentials['access_token']
+
+
+def process(domain, client_id, client_secret, authz_api, app_name, email):
+    # create API client
+    auth0 = get_auth0_client(domain, client_id, client_secret)
+
+    # get authorization extension token
+    authz_token = get_authz_token(domain, client_id, client_secret)
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: I decided to get rid of `process()` altogether and have its step in the `main` function.

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    user = create_passwordless_user(auth0, email)
+    user_id = user['user_id']
+
+    # Get shiny app by name
+    app = get_app(auth0, app_name)
+    app_id = app['client_id']
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: I'm changing to raise more exceptions now.

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    permission_api = PermissionAPI(authz_api, authz_token)
+    permission = permission_api.create(app_id, PERMISSION_NAME)
+
+    role_api = RoleAPI(authz_api, authz_token)
+    role = role_api \
+        .create(app_id, ROLE_NAME) \
+        .add_permission(permission.id())
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: I'll remove the fluent interface 👍

File: jenkinsfiles/auth0-users/create_auth0_user.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    group_api = GroupAPI(authz_api, authz_token)
+    group = group_api \
+        .create(app_name) \
+        .add_role(role_id) \
+        .add_user(user_id)
+
+
+if __name__ == '__main__':
+    import sys
+    sys.exit(main())
```
Comment (xoen) [inline jenkinsfiles/auth0-users/create_auth0_user.py]: 👍

File: jenkinsfiles/auth0-users/group.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+logging.basicConfig()
+LOG = logging.getLogger(__name__)
+LOG.setLevel(logging.DEBUG)
+
+
+class Group(object):
+
+    def __init__(self, authz_api, authz_token, group_data):
+        if not group_data:
+            raise ValueError("group_data can't be empty")
```
Comment (xoen) [inline jenkinsfiles/auth0-users/group.py]: Thanks, I'm now using `raise Exception(msg)`.

File: jenkinsfiles/auth0-users/group.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            json=[role_id],
+        )
+        if resp.status_code != 204:
+            LOG.error("Failed to add role ({}) to group ({}): expected 204, got {}: {}".format(role_id, self.id(), resp.status_code, resp.text))
+            return None
+
+        LOG.debug("Role ({}) added to group ({})".format(role_id, self.id()))
+        return self
+
+
```
Comment (xoen) [inline jenkinsfiles/auth0-users/group.py]: I've PEP8-ed everything ;)

File: jenkinsfiles/auth0-users/group.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                "Authorization": "Bearer {}".format(self.authz_token)
+            },
+            json=[user_id],
+        )
+        if resp.status_code != 204:
+            LOG.error("Failed to add user ({}) to group ({}): expected 204, got {}: {}".format(user_id, self.id(), resp.status_code, resp.text))
+            return None
+
+        LOG.debug("User ({}) added to group ({})".format(user_id, self.id()))
+        return self
```
Comment (xoen) [inline jenkinsfiles/auth0-users/group.py]: I'll see what I can do

File: jenkinsfiles/auth0-users/group_api.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+              'description': group_name,
+            }
+        )
+        if resp.status_code != 200:
+            LOG.error("Failed to create group: expected 200, got {}: {}".format(resp.status_code, resp.text))
+            return None
+        group = resp.json()
+
+        LOG.debug("Group created = {}".format(group))
+        return group
```
Comment (xoen) [inline jenkinsfiles/auth0-users/group_api.py]: I sorted this.

File: jenkinsfiles/auth0-users/permission.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import requests
```
Comment (xoen) [inline jenkinsfiles/auth0-users/permission.py]: 👍

File: jenkinsfiles/auth0-users/role.py (hunk 1 above)
Comment (xoen) [inline jenkinsfiles/auth0-users/role.py]: I just realised why I can't, it's because the API will complain about other attributes.

**Comment by xoen** [comment]
Hey @andyhd here I applied your suggestions: https://github.com/ministryofjustice/analytics-platform-ops/pull/35/commits/ccef09234c6f3608d2586b679382948378369f79

**Comment by xoen** [comment]
FYI: Rebased off `master` and merging as discussed offline.

**Pull request merged**